File: CS/CustomAutoFilterRowFiltering/Custom Grid.cs
using System;
using System.Collections.Generic;
using DevExpress.XtraGrid;
using DevExpress.XtraGrid.Registrator;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.Data.Filtering;
using DevExpress.XtraGrid.Columns;
using System.Text;
using System.Globalization;
using System.ComponentModel;


namespace DXSample {
    [ToolboxItemAttribute(true)]
    public class MyGridControl : GridControl
    {
        protected override BaseView CreateDefaultView()
        {
            return CreateView("MyGridView");
        }
        protected override void RegisterAvailableViewsCore(InfoCollection collection)
        {
            base.RegisterAvailableViewsCore(collection);
            collection.Add(new MyGridViewInfoRegistrator());
        }
    }
    public class MyGridViewInfoRegistrator : GridInfoRegistrator
    {
        public override string ViewName { get { return "MyGridView"; } }
        public override BaseView CreateView(GridControl grid) { return new MyGridView(grid as GridControl); }
    }
    public class MyGridView : GridView
    {
        public MyGridView() : this(null) { }
        public MyGridView(DevExpress.XtraGrid.GridControl grid) : base(grid) { filteredColumns = new Dictionary<GridColumn, ColumnFilterInfo>(); }

        protected override string ViewName { get { return "MyGridView"; } }

        Dictionary<GridColumn,ColumnFilterInfo> filteredColumns = null;
        protected override DevExpress.Data.Filtering.CriteriaOperator CreateAutoFilterCriterion(DevExpress.XtraGrid.Columns.GridColumn column, DevExpress.XtraGrid.Columns.AutoFilterCondition condition, object _value, string strVal)
        {

            CriteriaOperator op = base.CreateAutoFilterCriterion(column, condition, _value, strVal);
            if (strVal.Length > 0)
            {
                if (!filteredColumns.ContainsKey(column))
                    filteredColumns.Add(column, new ColumnFilterInfo(ColumnFilterType.AutoFilter, _value, op, string.Empty));
                else
                    filteredColumns[column] = new ColumnFilterInfo(ColumnFilterType.AutoFilter, _value, op, string.Empty);
                return op;
            }
            return op;
        }


 
        public static string RemoveDiacritics(string src, bool compatNorm)
        {
            StringBuilder sb = new StringBuilder();
            foreach (char c in src.Normalize(compatNorm ? NormalizationForm.FormKD : NormalizationForm.FormD))
                switch (CharUnicodeInfo.GetUnicodeCategory(c))
                {
                    case UnicodeCategory.NonSpacingMark:
                    case UnicodeCategory.SpacingCombiningMark:
                    case UnicodeCategory.EnclosingMark:
                        //do nothing
                        break;
                    default:
                        sb.Append(c);
                        break;
                }
            
            return sb.ToString();
        }
 

   
        enum FilterResult{ NotProcessed = -1, RowNotExists = 0, RowExists};
        
        protected override int RaiseCustomRowFilter(int listSourceRow, bool fit)
        {
            int result = base.RaiseCustomRowFilter(listSourceRow,fit);
            bool IsCustomRowFilterEventHandled = result != -1;
            if (IsCustomRowFilterEventHandled) return result;

            string filterString = string.Empty;
            DevExpress.XtraGrid.Columns.AutoFilterCondition operandType = AutoFilterCondition.Default;
            GridColumn filteredColumn = null;

            FilterResult filterResult = FilterResult.RowExists;
            foreach (KeyValuePair<GridColumn, ColumnFilterInfo> filterPair in filteredColumns)
            {
                ColumnFilterInfo filterInfo = filterPair.Value;
                if (filterInfo.Type != ColumnFilterType.AutoFilter)
                    return (int)FilterResult.NotProcessed;
                filteredColumn = filterPair.Key;
                if (filteredColumn.ColumnType != typeof(string))
                {
                    filterResult = FilterResult.NotProcessed;
                    continue;
                }

                operandType = filteredColumn.OptionsFilter.AutoFilterCondition;
                filterString = filterInfo.Value.ToString();

                string value = this.GetRowCellDisplayText(listSourceRow, filteredColumn);
                if (value == string.Empty)
                {
                    filterResult = FilterResult.NotProcessed;
                    continue;
                }
                string processedString = RemoveDiacritics(value, true).ToLower();
                string filter = RemoveDiacritics(filterString, true).ToLower();
                if (operandType == AutoFilterCondition.Like || operandType == AutoFilterCondition.Default)
                {
                    if (!processedString.StartsWith(filter))
                    {
                        filterResult = FilterResult.RowNotExists;
                    }
                }
                if (operandType == AutoFilterCondition.Contains)
                {
                    if (!processedString.Contains(filter))
                    {
                        filterResult = FilterResult.RowNotExists;
                    }
                }
                if (operandType == AutoFilterCondition.Equals)
                {
                    if (!processedString.Equals(filter))
                    {
                        filterResult = FilterResult.RowNotExists;
                    }
                }

            }

            return (int)filterResult;
        
        }

        protected override void RaiseColumnFilterChanged()
        {
            base.RaiseColumnFilterChanged();
            List<GridColumn> columnsToRemove = new List<GridColumn>();
            foreach (KeyValuePair<GridColumn, ColumnFilterInfo> filterPair in filteredColumns)
            {
                GridColumn column = filterPair.Key;
                if (column.FilterInfo.FilterString == string.Empty)
                {
                    columnsToRemove.Add(column);
                }
            }

            for (int i = 0; i < columnsToRemove.Count; i++)
            {
                if (filteredColumns.ContainsKey(columnsToRemove[i]))
                    filteredColumns.Remove(columnsToRemove[i]);
            }
            columnsToRemove.Clear();
            this.RefreshData();
        } 
 

          
    }
}

File: VB/CustomAutoFilterRowFiltering/Custom Grid.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports DevExpress.XtraGrid
Imports DevExpress.XtraGrid.Registrator
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.Data.Filtering
Imports DevExpress.XtraGrid.Columns
Imports System.Text
Imports System.Globalization
Imports System.ComponentModel


Namespace DXSample
	<ToolboxItemAttribute(True)> _
	Public Class MyGridControl
		Inherits GridControl
		Protected Overrides Function CreateDefaultView() As BaseView
			Return CreateView("MyGridView")
		End Function
		Protected Overrides Sub RegisterAvailableViewsCore(ByVal collection As InfoCollection)
			MyBase.RegisterAvailableViewsCore(collection)
			collection.Add(New MyGridViewInfoRegistrator())
		End Sub
	End Class
	Public Class MyGridViewInfoRegistrator
		Inherits GridInfoRegistrator
		Public Overrides ReadOnly Property ViewName() As String
			Get
				Return "MyGridView"
			End Get
		End Property
		Public Overrides Function CreateView(ByVal grid As GridControl) As BaseView
			Return New MyGridView(TryCast(grid, GridControl))
		End Function
	End Class
	Public Class MyGridView
		Inherits GridView
		Public Sub New()
			Me.New(Nothing)
		End Sub
		Public Sub New(ByVal grid As DevExpress.XtraGrid.GridControl)
			MyBase.New(grid)
		filteredColumns = New Dictionary(Of GridColumn, ColumnFilterInfo)()
		End Sub

		Protected Overrides ReadOnly Property ViewName() As String
			Get
				Return "MyGridView"
			End Get
		End Property

		Private filteredColumns As Dictionary(Of GridColumn,ColumnFilterInfo) = Nothing
		Protected Overrides Function CreateAutoFilterCriterion(ByVal column As DevExpress.XtraGrid.Columns.GridColumn, ByVal condition As DevExpress.XtraGrid.Columns.AutoFilterCondition, ByVal _value As Object, ByVal strVal As String) As DevExpress.Data.Filtering.CriteriaOperator

			Dim op As CriteriaOperator = MyBase.CreateAutoFilterCriterion(column, condition, _value, strVal)
			If strVal.Length > 0 Then
				If (Not filteredColumns.ContainsKey(column)) Then
					filteredColumns.Add(column, New ColumnFilterInfo(ColumnFilterType.AutoFilter, _value, op, String.Empty))
				Else
					filteredColumns(column) = New ColumnFilterInfo(ColumnFilterType.AutoFilter, _value, op, String.Empty)
				End If
				Return op
			End If
			Return op
		End Function



		Public Shared Function RemoveDiacritics(ByVal src As String, ByVal compatNorm As Boolean) As String
			Dim sb As New StringBuilder()
			For Each c As Char In src.Normalize(If(compatNorm, NormalizationForm.FormKD, NormalizationForm.FormD))
				Select Case CharUnicodeInfo.GetUnicodeCategory(c)
					Case UnicodeCategory.NonSpacingMark, UnicodeCategory.SpacingCombiningMark, UnicodeCategory.EnclosingMark
						'do nothing
					Case Else
						sb.Append(c)
				End Select
			Next c

			Return sb.ToString()
		End Function



		Private Enum FilterResult
			NotProcessed = -1
			RowNotExists = 0
			RowExists
		End Enum
		Protected Overrides Function RaiseCustomRowFilter(ByVal listSourceRow As Integer) As Integer
			Dim result As Integer = MyBase.RaiseCustomRowFilter(listSourceRow)
			Dim IsCustomRowFilterEventHandled As Boolean = result <> -1
			If IsCustomRowFilterEventHandled Then
				Return result
			End If

			Dim filterString As String = String.Empty
			Dim operandType As DevExpress.XtraGrid.Columns.AutoFilterCondition = AutoFilterCondition.Default
			Dim filteredColumn As GridColumn = Nothing

			Dim filterResult As FilterResult = FilterResult.RowExists
			For Each filterPair As KeyValuePair(Of GridColumn, ColumnFilterInfo) In filteredColumns
				Dim filterInfo As ColumnFilterInfo = filterPair.Value
				If filterInfo.Type <> ColumnFilterType.AutoFilter Then
					Return CInt(Fix(FilterResult.NotProcessed))
				End If
				filteredColumn = filterPair.Key
				If filteredColumn.ColumnType IsNot GetType(String) Then
					filterResult = FilterResult.NotProcessed
					Continue For
				End If

				operandType = filteredColumn.OptionsFilter.AutoFilterCondition
				filterString = filterInfo.Value.ToString()

				Dim value As String = Me.GetRowCellDisplayText(listSourceRow, filteredColumn)
				If value = String.Empty Then
					filterResult = FilterResult.NotProcessed
					Continue For
				End If
				Dim processedString As String = RemoveDiacritics(value, True).ToLower()
				Dim filter As String = RemoveDiacritics(filterString, True).ToLower()
                If operandType = AutoFilterCondition.Like OrElse operandType = AutoFilterCondition.Default Then
                    If (Not processedString.StartsWith(filter)) Then
                        filterResult = filterResult.RowNotExists
                    End If
                End If
                If operandType = AutoFilterCondition.Contains Then
                    If (Not processedString.Contains(filter)) Then
                        filterResult = filterResult.RowNotExists
                    End If
                End If
                If operandType = AutoFilterCondition.Equals Then
                    If (Not processedString.Equals(filter)) Then
                        filterResult = filterResult.RowNotExists
                    End If
                End If

			Next filterPair

			Return CInt(Fix(filterResult))

		End Function

		Protected Overrides Sub RaiseColumnFilterChanged()
			MyBase.RaiseColumnFilterChanged()
			Dim columnsToRemove As New List(Of GridColumn)()
			For Each filterPair As KeyValuePair(Of GridColumn, ColumnFilterInfo) In filteredColumns
				Dim column As GridColumn = filterPair.Key
				If column.FilterInfo.FilterString = String.Empty Then
					columnsToRemove.Add(column)
				End If
			Next filterPair

			For i As Integer = 0 To columnsToRemove.Count - 1
				If filteredColumns.ContainsKey(columnsToRemove(i)) Then
					filteredColumns.Remove(columnsToRemove(i))
				End If
			Next i
			columnsToRemove.Clear()
			Me.RefreshData()
		End Sub



	End Class
End Namespace